### ISSUE TITLE
Fotorama product picture script is extremely slow on mobile devices

### ISSUE THREAD

**Issue by andidhouse** [magento/magento2#6018]
1. Magento 2.1 via download / nginx / php7 / redis / varnish / mariadb
### Steps to reproduce
1. Go to a product page with a smartphone (even if no product picture is uploaded and the default backup picture is there)
2. look at the load time of the product page
### Expected result
1. The product site loads very quick on smartphones and tablets too
### Actual result
1. The fotorama script is extremely slow on smartphones and causes the whole product site to load several seconds without showing a picture.
   You will realize that that all parts of the product page load extreme fast but not the fotorama picture script. 
   Also if there is no picture uploaded and the default backup picture is loaded (with just a view kbs) the script needs seconds to render the default backup picture with just a view kbs.

I think this is not exactable on mobile devices at all. There are other ways to load the product pictures first before calling a complex script which blocks the load of the first picture and just shows a spinner for several seconds.

**Comment by andidhouse** [comment]
We also checked this on multi servers and installations - it is not the server power or anything else.
It is the fotorama script or the implementation of it. 

You can for example check out this magento2 site and you will see that the product sites need several seconds to load the pictures on mobile devices:

https://www.ninewest.co.uk

**Comment by piotrekkaminski** [comment]
Thank you for your submission.

We recently made some changes to the way we process GitHub submissions to more quickly identify and respond to core code issues.

Feature Requests and Improvements should now be submitted to the new [Magento 2 Feature Requests and Improvements](https://community.magento.com/t5/Magento-2-Feature-Requests-and/idb-p/feature-requests) forum (see details [here](https://community.magento.com/t5/News-Announcements/Improvements-to-GitHub-Management/m-p/44572#M96)).



We are closing this GitHub ticket and have moved your request to the new [forum](https://community.magento.com/t5/Magento-2-Feature-Requests-and/Fotorama-product-picture-script-is-extremely-slow-on-mobile/idi-p/45011).

**Comment by lucscu** [comment]
@piotrekkaminski How could you move huge bugs (it's incredible that you release a so slower carousel builder for mobile) to "Feature Requests and Improvements"?
How could a user waits 5 seconds before see product's image? And in the meantime see the ugliest loader ever?

**Comment by choukalos** [comment]
Reopening; changing to bug.  Internal ticket - MAGETWO-58628

Hi @andidhouse & @LucScu ; sorry our script that runs under Piotr's name auto-moved this.  Looks like it was incorrectly tagged with the label improvement.  

I did notice that JS on the site hasn't been bundled & minified.  That should help performance some.

**Comment by piotrekkaminski** [comment]
@LucScu we have automated process that moves every ticket markes as improvement or feature request to forums. I agree this should be fixed so we have reopened it. Thanks!

**Comment by lucscu** [comment]
@choukalos On my env all js and css are bundled & minified, but product's pages are very bad user experience due to slowness of fotorama js. 
Are you agree?
Do you have a fix to correct it or we have to install a 3rd part extension?

**Comment by andidhouse** [comment]
@choukalos The bundling and minify is another topic here. As far as i know this does not work correctly and is reported several times here. 

As @LucScu states here - also if we bundle and minify js and css this has no huge effect. The loading time of the script is simply inacceptable for a mobile or tablet user.

**Comment by ignaciomella** [comment]
I cannot stand this. Product images loads too slow. 

There's a temporary solution posted in StackExchange 

[http://magento.stackexchange.com/questions/134834/how-can-we-load-product-image-gallery-instantly-without-the-delay-in-magento](http://magento.stackexchange.com/questions/134834/how-can-we-load-product-image-gallery-instantly-without-the-delay-in-magento)

From what I read there, this problem lies with Magento's implemenation of the fotorama script.

**Comment by anton-huz** [comment]
@andidhouse ,
You pointed that Fotorama (and so Magento's Gallery.js build on top of Fotorama) works slow. It is wrong opinion. Gallery itself work fast enough (as any other image gallery widget could).

## Why not a Fotorama?

I've made some changes to PDP to leave just Gallery with dependencies and compare page speed:

| Instance | Total requests | Total data | Time of page processing | Time of page processing on low connection |
| --- | --- | --- | --- | --- |
| Default | 179 | 882Kb | [2sek](https://cloud.githubusercontent.com/assets/1562842/18993961/6f777394-872e-11e6-9fd0-a92baeae734c.png) | [46sek](https://cloud.githubusercontent.com/assets/1562842/18993960/6f7699d8-872e-11e6-962d-14cbf3a76129.png) |
| Gallery only | 15 | 218Kb | [0.12sek](https://cloud.githubusercontent.com/assets/1562842/18993962/6f794750-872e-11e6-9bd2-9d0dcc27f666.png) | [1sek](https://cloud.githubusercontent.com/assets/1562842/18993963/6f7d6970-872e-11e6-9af3-be487648b7e7.png) |

Instance with cut PDP could be set up from [my test branch](https://github.com/guz-anton/magento2/tree/MAGETWO-58628).

## Who is if not a Fotorama?

Main admonitions are 
- Visual appearance of images is too late (like User feeling of "fast")
- Magento JS app on storefront has too long intializization (it's too huge)

First point could be solved. Example by [Aaron Allen](http://magento.stackexchange.com/q/134834/33166) from previous comment. Yes, adding the image on stage could do such trick.

Second point is a subject of set of improvements. I don't think that we should track them under this Github issue. So I'm going to close an issue like _Won't fix_.

Please leave notes here if you have additional insights.

**Comment by andidhouse** [comment]
@guz-anton can you pls point out a live magento2 website where we can test and confirm this. As far as i know the script is the problem and really slow on mobile - a killer.

I am happy if you prove me wrong here but pls point out a m2 site where we can test this.

Not shure about your point here:
Default 179 882Kb   2sek    46sek
Gallery only    15  218Kb   0.12sek 1sek

To be honest i do not believe that a default call of a site with pictures should last 46 seconds on a low internet connection. Pls also have a look at the quality. 218kb is not 882kb so the question is really how photorama is compressing the images. Do you have any infos here?

Many thanks!

**Comment by andidhouse** [comment]
@guz-anton can you pls read my post carefully:

--> IT IS ABOUT smartphones and tablets 

The normal website view is ok. But try to use a smartphone to view the pictures. It loads not in 1-2 seconds but takes 3-6 seconds on a very fast server and internet connection.

**Comment by anton-huz** [comment]
@andidhouse 
I understand that original issue reported for mobile devices. I have no comments that mobile performance is worse than desktop ([but in nowadays it could be an opposite](http://www.techugly.com/2016/09/iphone-7-is-faster-than-any-macbook-pro.html)). Sorry, that screenshots show picture only for the desktop.

I have concerns about _Fotorama_ word in the issue description. We cannot fix just _Fotorama_ 's performance. Whole picture with JavaScript app on storefront should be improved. We will continue work in this direction.

**Comment by andidhouse** [comment]
@guz-anton many thanks for your reply.
It is possible that your send me some links of M2 live sites where you think the performance ist ok. My team will then check this.

For example we did a test using magic zoom extension instead of fotorama - speed was more then 5 times faster on mobile devices with magic zoom.

**Comment by anton-huz** [comment]
Thanks for point about comparison _magic zoom_ vs _fotorama_ . I'll check. As about instance - ping me on aguz@magento.com .

**Comment by lucscu** [comment]
@andidhouse How do you implement magic zoom? Do you install a 3rd part extension or manually?

**Comment by ghost** [comment]
@LucScu whatever works for your budget.

**Comment by lucscu** [comment]
@jupiter01 do you have a magic zoom deployed on a demo site?

**Comment by ghost** [comment]
No, a quick google will bring up the magic zoom website, they have live demos there.

**Comment by andidhouse** [comment]
You can download the M2 demo of magic zoom on the website - install it in 2 minutes and test it.
Can you pls share your results on mobile here?

**Comment by lucscu** [comment]
@jupiter01 ok
@andidhouse for now i'll continue with default fotorama, but when i'll test magic zoom i'll give here a feedback

**Comment by andidhouse** [comment]
@choukalos really don´t know why this is closed without a solution... it is also reported in the forum:

http://magento.stackexchange.com/questions/134834/how-can-we-load-product-image-gallery-instantly-without-the-delay-in-magento

**Comment by lucscu** [comment]
@andidhouse try to use magic zoom, it is easy to install (2 minutes and free demo) and it works like a charm!

**Comment by andidhouse** [comment]
@LucScu many thanks. yes we tried this and it works great.
But we try to use as less extensions for normal functions as possible due to update problems and conflicts. I think it is a very poor implementation of the magento gallery script at the moment causing so much trouble.... i am not shure they fix it the next time so we maybe give magic zoom a try... :-/

**Comment by andidhouse** [comment]
we give it up on the native photorama implementation due to:

- problem with mobile devices and scrolling issue
- extreme long load times of the script compared to for example magic zoom
- some other small bugs and in general poor implementation of the script at this moment

Magic zoom works like a charm and has a great support fixing bugs in minute not years.

**Comment by andidhouse** [comment]
just tested in 2.1.3 against magic zoom.

Problem is present - on other magento2 live sites as well.
Problem with implementation of fotorama in m2 at this stage:

The first product picture is shown AFTER all pictures are downloaded meaning if you have 2-5 pictures on one product this takes about 3-10 seconds on mobile devices and about 1-2 seconds more on desktop.

This behaviour is very bad for user experience as well as for google. I am not shure why this problem is not taken into account here!

**Comment by Crossmotion** [comment]
I have the same isseu on version 2.1.3. desktop version load under 1 sec. mobile version shows a loading icon for about 11 seconds before showing image ( there is only one image in this product)

**Update:** I first tested it on a 2 year old HTC M8. But it seems to work better on newer phones (tested on Iphone 7 and Nexus 6P). So it looks like its mostly an performance issue on older phones. but also on new phones you can see a delay.

**Comment by andidhouse** [comment]
@Crossmotion exactly - it is a poor implementation of the fotorama script performance wise. On newer  and faster systems you can hardly see a delay. if you turn to some older devices there is a big delay in loading the page. If you use for example magic zoom extension there is no delay at all so this is clearly a bad implementation of the fotorama script.

**Comment by twistedatrocity** [comment]
@andidhouse I"m having the same problem. There is a noticeable delay using a desktop, like a good 2-3 seconds if images are in the cache, longer if not, and this is on a beastly modern gaming laptop. However on a mobile, Galaxy S4 with chrome, the delay is a solid 10 seconds at times, more if images not cached. The phone is dated but isn't THAT old. The phone still works well on other ecommerce sites, like amazon for example.

The performance is also poor once the images are loaded. Tapping and zooming on mobile is noticeable twitchy/lagginess. The close button when zooming is also right over the cart icon (in luma theme), when tapping close, the tap bleeds through the element and also activates the cart link. 

This is way too long of delays for E commerce, customers will go elsewhere. All we have are images to sell a product. Product descriptions mean nothing without a picture.

Did you ever find a solution? I hate to buy yet another extension to address core functionality.

**Comment by andidhouse** [comment]
@twistedatrocity no solution at all - magento team just does "part fixing" like the picture scroll bug. I suggest two ways:

- install magic zoom plus which works lightning fast and solid
- get rid of magento2 and change to another system ;-)

You can not trust magento to at this stage to implement bugfixes as fast as needed and also as solid as needed.

**Comment by andidhouse** [comment]
@piotrekkaminski are you really shure it is a good idea to close such a major not solved mug here at all? No words on this...

**Comment by ignaciomella** [comment]
I'm back on magento1 ... Current Magneto 2 commitment falls short of many of us who expected great things. To date.. I don't see the point of upgrading to M2... lots of problems (like this thread) and not any major benefits. Sad.

**Comment by TimQSO** [comment]
@guz-anton You say:

'I have concerns about Fotorama word in the issue description. We cannot fix just Fotorama 's performance. Whole picture with JavaScript app on storefront should be improved. We will continue work in this direction.'

I doubt this is because of the performance of Fotorama. When I look at the waterfall of a product page of our staging server, I can see the following:

1. Total requests: 187
2. Cross sell and up sell product images loaded in request 9 to 13
3. Fotorama.js loaded in request 166
4. Actual image of the product loaded in request **184**

This has nothing to do with the performance of Fotorama itself. As I can see dozens of examples of fast fotorama implementations. This has something to do with incorrect implementation of the fotorama software in Magento 2 and nothing else.

It worries me that several people complain about it, that the issue is very easy to reproduce and that Magento developers just seem to ignore this issue. While it is a very big performance issue.

Also on fast desktops the delay is noticable and a very annoying thing in user experience. It is horrible on mobile phones.

Look at this example, this is a very fast Magento 2 demoserver with PHP7, Redis and Varnish:

https://varnish.m2.magehost.pro/stellar-solar-jacket

You must agree this performance is just a bit sad, annoying and incomprehensible, even on a desktop!

I really don't care if this should be considered as a bug, missing feature, misimplementation, optimization or whatever. The fact is that Magento 2 store owners are having severe performance issues with displaying product images for crying out loud and Magento just doesn't fix this issue.

What is wrong with this picture? (Well this is a nice word joke...)

**Comment by andidhouse** [comment]
@andimov pls have a look on this again...  i really do not know why magento is ignoring a major bug like this since month!

**Comment by TimQSO** [comment]
I really think that the main issue is that the Fotorama and the Gallery parts are loaded way too late!

I mean, loading the actual product photo as request 184 while there are 187 requests in total? It is almost the latest request to be loaded. On a product page a product photo should be loaded in the top 10 of requests.

It even comes after the 'Document Complete', Webpagetest.org tells met 'Document Complete' after request 101. 

You can see my example here:

http://www.webpagetest.org/result/170320_AW_aa44cf01d093e195a188d8d5d263cae8/

**Comment by TimQSO** [comment]
It is disappointing that there is no response or solution whatsoever. 22 days ago I tried to comment on this topic and I'm just being ignored.

The question is: why does Magento does not take this issue seriously? I think it is a severe issue that must have been solved a year ago. Instead this is being ignored .

Tim

**Comment by ngz82** [comment]
+1 for fixing this. It is a terrible user experience on mobile.

**Comment by ghost** [comment]
I can't believe this is still an issue.

**Comment by ngz82** [comment]
I've just noticed this same delay of the product images, seems to impact the loading of configurable product "options" in the drop down menu on the product page. The configurable options only show up once the images load so that could be linked.  
This effectively means a customer cannot add these products to their shopping cart until it all loads, and in some cases this is many seconds -  a very poor experience.

**Comment by johnny-longneck** [comment]
I had the same problem and found out it was Paypal Express that blocked the loading of the images and other things on the page. 
In the backend there is a setting for the PayPal Express Button. Don't set it to "dynamic" -> Use "static". 
Otherwise it makes a request to paypal and this is causing the problem.

At least it solved the problem for me.

**Comment by TimQSO** [comment]
Hmm I tried this immediately, but no effect for my site unfortunately.

I timed it and it is as slow as ever. I also don't use that Paypal plugin at all,  but still changed it do static, without any effect :(

My site has about 200 requests on one page. And the actual product image is still somewhere down the row, request 174.

**Comment by yumicom** [comment]
I"m having the same problem

**Comment by moderncodes** [comment]
### Preconditions
1. Magento 2.2.0
2. PHP 7.0.24
3. NGINX 1.9.14 PHP-FPM
4. Redis session/page caching
5. https
6. AVG site speed load time around 1-1/15 seconds

### Steps to reproduce

1. Load on desktop browser simple or configurable product page

### Expected result
1. Page load within 1-2 seconds

### Actual result
1. Page freezes and becomes unresponsive. User unable to click, scroll or anything else, except closing the tab.
2. Main product image placeholder having a spinning wheel.
3. In developer console "Network" tab, all loading product images locked to status "pending".
All until async google-analytics.com/plugins/ua/ec.js completes request response, which on average anywhere between 3-10 seconds.
With this unfortunate issue, all the site speed optimization becomes irrelevant.

Removing google analytics tracker, fixes the problem, and of course creates a new one, I get no analytics.

![example-screenshot](https://user-images.githubusercontent.com/12808762/31697928-7337cfe2-b388-11e7-9305-17e003330f88.jpg)

Can someone please help? I hate to waste so much effort to bring site speed up, and then just to quit on because of something like this.

**Comment by qrider** [comment]
We are experiencing very similar issues on our site for the products that have youtube videos. I am using a Macbook air and seeing extremely long load times.

**Comment by tomekjordan** [comment]
Hi. tested M211 and this is still the main problem from UX perspective. Maybe automated tests shows it load fast, but what users sees is different. From my perspective M1 merchant. This is top prio issue to fix. I cannot move to M2 until this slow image + ugly loader will not be improved.

The common problem is that even 3rd party themes comes with THIS "feature" :)
If user want to see product page or open cart page - it takes too much time to load, even on fast server.  Because images and ajaxed blocks defer loading is breaking UX for customers, I expect conversion on M2 site will be lower than on M1.
Since this is ecommerce - conversion & UX for user is TOP1 factor for any business

Hope somebody will find solution and fix the core with it :)

**Comment by TimQSO** [comment]
Why is this issue marked as closed? It is a bit weird that such an issue is being ignored, isn't it?

**Comment by yumicom** [comment]
I closed this issue - was forced to buy a third-party program to speed up the opening of fotorama product pictures. Thank you, Magento team ((

**Comment by TimQSO** [comment]
Which one did you buy @yumicom ?

**Comment by yumicom** [comment]
I bought Magic Zoom (not Plus).

**Comment by TimQSO** [comment]
Ok, that one was on my list too. Thanks!

**Comment by royandre** [comment]
We are also writing our own image magnifier as the one bundled in Magento (even Magento Commerce) is simply useless due to this.

Need to cast a -1 to the Magento dev team on this one.

Also worth mentioning that fotorama has been deprecated (https://github.com/artpolikarpov/fotorama/issues/532)

**Comment by harrigo** [comment]
I'm having similar issues with this script just seems so slow however not just on mobiles. It can take 3 or 4 seconds to load on desktops especially chrome. This feels really slow as the rest of the page is up and running in 0.2 seconds. Am debating ripping it out for Magic Zoom if magic zoom works with all the swatches etc.

This however does seem intermittent on desktops as a cache flush seems to being the speed down however have seen computers taking minutes on some occasions to load images until I flushed cache. Not sure if this is a separate issue however.

**Comment by siliconalchemy** [comment]
Incredible that this is still a problem, and this issue remains closed.  This is a major issue on all browsers including desktops.  In my case the page loads <1-2sec, then it takes anything up to 3-10sec (depending on the speed of the device/browser) to display the main picture - just a spinner in the meantime.  Terrible, terrible UX.

**Comment by tomekjordan** [comment]
100% agree. terrible UX

**Comment by ngz82** [comment]
Terrible UX indeed. We went ahead and installed magic zoom, problem solved.

**Comment by DrewML** [comment]
Related: https://github.com/magento/magento2/issues/14667

**Comment by DrewML** [comment]
For those still following this thread, we're discussing alternatives to address the performance issue https://github.com/magento/architecture/pull/33

**Comment by LiamKarlMitchell** [comment]
Still slow.

**Comment by ParimalSigma** [comment]
We are also experiencing the same issue with latest Magento version 2.4.2

**Comment by an4rei** [comment]
5 years later... same issue

**Comment by GuiltyNL** [comment]
Just buy Magic Zoom guys. It is really not worth waiting on fixes. Adobes priority clearly lies somewhere else.

https://www.magictoolbox.com/magiczoom/modules/magento/
